src/app/reporting/Engine.cpp

```diff
@@ -449,12 +459,14 @@ void Engine::Run()
     uint32_t numReadHandled = 0;
 
     InteractionModelEngine * imEngine = InteractionModelEngine::GetInstance();
-    ReadHandler * readHandler         = imEngine->mReadHandlers + mCurReadHandlerIdx;
 
     mRunScheduled = false;
 
-    while ((mNumReportsInFlight < CHIP_IM_MAX_REPORTS_IN_FLIGHT) && (numReadHandled < CHIP_IM_MAX_NUM_READ_HANDLER))
+    while ((mNumReportsInFlight < CHIP_IM_MAX_REPORTS_IN_FLIGHT) && (numReadHandled < imEngine->mReadHandlers.Allocated()))
     {
+        ReadHandler * readHandler = imEngine->mReadHandlers.At(mCurReadHandlerIdx % imEngine->mReadHandlers.Allocated());
```

Comment:
If I understand correctly, this code is maintaining mCurReadHandlerIdx in order to enforce fairness between all read handlers in accessing the available pool of CHIP_IM_MAX_REPORTS_IN_FLIGHT during each Run() call.

However, the O(N^2) heap implementation `At()` inside this loop and the need to reset mCurReadHandlerIdx because indices aren't stable are showing that this is an awkward approach (and also not fair if mCurReadHandlerIdx is reset frequently).  I also suspect this may break for the bitmap implementation, where I think At(idx) will return inactive objects.

Another (better?) way to solve this fairness problem is for reporting requests to enqueue.  If read handlers with readHandler->IsReportable() == true enqueue when IsReportable is set, you can simply pick handlers off the head of the queue until you exhaust CHIP_IM_MAX_REPORTS_IN_FLIGHT.  The queue should be implementable within the ReadHandler object.

Reply:
> where I think At(idx) will return inactive objects.

Good catch! I completely missed that - unfortunately, the documentation there isn't all that stellar, but nonetheless, I had incorrectly assumed that `At` returned a positional index into the active array.

I'll need to think on this a bit, and mull over your suggestion as well.